quickwit-storage/src/cache/storage_with_cache.rs

```diff
@@ -0,0 +1,111 @@
+//  Quickwit
+//  Copyright (C) 2021 Quickwit Inc.
+//
+//  Quickwit is offered under the AGPL v3.0 and as commercial software.
+//  For commercial licensing, contact us at hello@quickwit.io.
+//
+//  AGPL:
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Affero General Public License as
+//  published by the Free Software Foundation, either version 3 of the
+//  License, or (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Affero General Public License for more details.
+//
+//  You should have received a copy of the GNU Affero General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use std::ops::Range;
+use std::path::Path;
+use std::sync::Arc;
+
+use crate::cache::Cache;
+use crate::{PutPayload, Storage, StorageFactory, StorageResult};
+
+struct StorageWithCache {
+    storage: Arc<dyn Storage>,
+    cache: Arc<dyn Cache>,
+}
+
+#[async_trait]
+impl Storage for StorageWithCache {
+    async fn put(&self, path: &Path, payload: PutPayload) -> StorageResult<()> {
+        // We actually do not cache anything on put as it is meaningless in Quickwit.
+        self.storage.put(path, payload).await
+    }
+
+    async fn copy_to_file(&self, path: &Path, output_path: &Path) -> StorageResult<()> {
+        if let Some(bytes) = self.cache.get_all(path).await {
```

Comment:
I guess this implies that the cache if it contains an entire local file, then it's always in sync with the disk version?

Reply:
Very good point. I changed the implementation to panic on write operations.